File: backend/src/models/SearchHistory.js
const db = require('../config/database');

class SearchHistory {
  static async add(userId, searchQuery, filters = {}, resultsCount = 0) {
    const query = `
      INSERT INTO search_history (user_id, search_query, filters, results_count)
      VALUES ($1, $2, $3, $4)
      RETURNING *
    `;
    const result = await db.query(query, [userId, searchQuery, JSON.stringify(filters), resultsCount]);
    return result.rows[0];
  }

  static async getByUserId(userId, limit = 20) {
    const query = `
      SELECT * FROM search_history
      WHERE user_id = $1
      ORDER BY created_at DESC
      LIMIT $2
    `;
    const result = await db.query(query, [userId, limit]);
    return result.rows;
  }

  static async clearHistory(userId) {
    const query = 'DELETE FROM search_history WHERE user_id = $1';
    await db.query(query, [userId]);
  }

  static async getPopularSearches(limit = 10) {
    const query = `
      SELECT search_query, COUNT(*) as search_count
      FROM search_history
      WHERE created_at > NOW() - INTERVAL '30 days'
      GROUP BY search_query
      ORDER BY search_count DESC
      LIMIT $1
    `;
    const result = await db.query(query, [limit]);
    return result.rows;
  }

  static async getSuggestionsFromHistory(userId, partialQuery, limit = 5) {
    const query = `
      SELECT DISTINCT search_query
      FROM search_history
      WHERE user_id = $1 AND search_query ILIKE $2
      ORDER BY created_at DESC
      LIMIT $3
    `;
    const result = await db.query(query, [userId, `${partialQuery}%`, limit]);
    return result.rows.map(row => row.search_query);
  }
}

module.exports = SearchHistory;

File: backend/app/models/search_history.py
import json
from typing import Any, Dict, List, Optional
from uuid import UUID

from app.config.database import database


class SearchHistory:
    @staticmethod
    async def add(
        user_id: UUID, 
        search_query: str, 
        filters: Optional[Dict[str, Any]] = None, 
        results_count: int = 0
    ) -> Optional[Dict[str, Any]]:
        query = """
            INSERT INTO search_history (user_id, search_query, filters, results_count)
            VALUES (:user_id, :search_query, :filters, :results_count)
            RETURNING *
        """
        result = await database.fetch_one(
            query=query, 
            values={
                "user_id": str(user_id), 
                "search_query": search_query, 
                "filters": json.dumps(filters) if filters else None, 
                "results_count": results_count
            }
        )
        return dict(result) if result else None

    @staticmethod
    async def get_by_user_id(user_id: UUID, limit: int = 20) -> List[Dict[str, Any]]:
        query = """
            SELECT * FROM search_history
            WHERE user_id = :user_id
            ORDER BY created_at DESC
            LIMIT :limit
        """
        results = await database.fetch_all(
            query=query, 
            values={"user_id": str(user_id), "limit": limit}
        )
        return [dict(r) for r in results]

    @staticmethod
    async def clear_history(user_id: UUID) -> None:
        query = "DELETE FROM search_history WHERE user_id = :user_id"
        await database.execute(query=query, values={"user_id": str(user_id)})

    @staticmethod
    async def get_popular_searches(limit: int = 10) -> List[Dict[str, Any]]:
        query = """
            SELECT search_query, COUNT(*) as search_count
            FROM search_history
            WHERE created_at > NOW() - INTERVAL '30 days'
            GROUP BY search_query
            ORDER BY search_count DESC
            LIMIT :limit
        """
        results = await database.fetch_all(query=query, values={"limit": limit})
        return [dict(r) for r in results]

    @staticmethod
    async def get_suggestions_from_history(
        user_id: UUID, 
        partial_query: str, 
        limit: int = 5
    ) -> List[str]:
        query = """
            SELECT DISTINCT search_query
            FROM search_history
            WHERE user_id = :user_id AND search_query ILIKE :partial_query
            ORDER BY created_at DESC
            LIMIT :limit
        """
        results = await database.fetch_all(
            query=query, 
            values={
                "user_id": str(user_id), 
                "partial_query": f"{partial_query}%", 
                "limit": limit
            }
        )
        return [r["search_query"] for r in results]
